src/runtime/wasm.ts

```diff
@@ -199,26 +251,29 @@ export class WasmParticle extends Particle {
   async initialize(buffer: ArrayBuffer) {
     assert(this.spec.name.length > 0);
 
-    // TODO: detect errors when this memory size doesn't match up with the wasm's declared values?
-    this.memory = new WebAssembly.Memory({initial: 256, maximum: 256});
-    this.heap = new Uint8Array(this.memory.buffer);
+    // TODO: vet the imports/exports on 'module'
+    const module = await WebAssembly.compile(buffer);
+    const emc = readEmscriptenMetadata(module);
+
+    this.memory = new WebAssembly.Memory({initial: emc.memSize, maximum: emc.memSize});
+    this.heapU8 = new Uint8Array(this.memory.buffer);
+    this.heap32 = new Int32Array(this.memory.buffer);
 
-    // TODO: find out how to calculate these values
-    const heap32 = new Int32Array(this.memory.buffer);
-    heap32[WasmConfig.DYNAMICTOP_PTR >> 2] = WasmConfig.DYNAMIC_BASE;
+    // We need to poke the address of the heap base into the memory buffer prior to instantiating.
+    this.heap32[emc.dynamictopPtr >> 2] = emc.dynamicBase;
 
     const env = {
       // Memory setup
       memory: this.memory,
-      __memory_base: WasmConfig.MEMORY_BASE,
-      table: new WebAssembly.Table({initial: 1, maximum: 1, element: 'anyfunc'}),
+      __memory_base: emc.globalBase,
+      table: new WebAssembly.Table({initial: emc.tableSize, maximum: emc.tableSize, element: 'anyfunc'}),
```

Comment:
This is the one I worry about. I think the way we can handle this, is if 'emc' isn't available in a custom section, we gracefully fall back to not including a table.  The current Kotlin bootstrap only includes only imported functions in the env, my guess is because the native wasm bootstrap creates its own table in the WASM code during compilation.

Reply:
Ok, so we could make emc create and pass the Table itself, and in the Kotlin case that's omitted.

Aside: I originally had a ts `type` EmConfig - hence `emc` - for the object returned by `readEmscriptenMetadata` but removed it because it just seemed redundant. Should I rename `emc` to `emConfig` for clarity?

Reply:
SGTM